### ISSUE TITLE
Fine grained control of the optimiser

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#1658]
Currently there are only 2 options:
- turn on/off optimisation
- number of constant optimiser runs

We do have a number of optimisations steps. It would be nice to have some of them switchable.

Eventually I would like to have a way to only run the peephole optimiser and nothing else if I wanted to.

**Comment by chriseth** [comment]
In the current version, there is no way to turn off peephole optimization (it runs even if you do not specify `--optimize`), so we should probably provide a switch for that.

**Comment by axic** [comment]
This can no be more easily achieved via json i/o.

**Comment by federicobond** [comment]
Does it make sense to define multiple optimization levels?

 * 0 would do no optimizations
 * 1 would do block deduplication and peephole optimizations
 * 2 would do all optimizations in level 1 plus constant optimizations

Note that many people get confused about the runs parameter, which does not define how many "passes" the constant optimizer does, but the expected number of times a constant will be resolved to a value once deployed. We may want to find a better name.

**Comment by axic** [comment]
I have a changeset which introduced flags for to turn on/off each optimiser step.

**Comment by chriseth** [comment]
Postponed to 0.5.0

**Comment by chriseth** [comment]
We should specify how to combine the IULIA optimizer settings with the legacy optimizer settings.

**Comment by axic** [comment]
To support yul optimisations steps some options were proposed today:
```
optimiser: [ “peephole”, “yul”, “constants:runs=200"…]
```

```
optimiser: {peephole: true, stack: true, ..., yul: true, yulOptions: { ...}}
```

where `yulOptions` could be a string `cse,ssa,join,inline(weight:8)`

**Comment by chriseth** [comment]
All options should allow some "default" settings.

**Comment by axic** [comment]
I'd like to keep support for "presets" aka what we have today, the only two presets: none vs. all optimisations.

**Comment by chriseth** [comment]
There should also be presets for the yul optimizer (we might not even implement the fine-grained control of the yul optimizer in the first release because the structure could change). Should "all" enable the yul optimizer or not? If not, can we change "all" to enable the yul optimizer at some point by default?

Another option that should be configurable: stack optimization in the code generator.

**Comment by chriseth** [comment]
note that also the binary search in the dispatcher depends on the number of expected executions.

**Comment by axic** [comment]
This was implemented via #5959.
